Refactor

CreepySkeleton opened the pull request:
<!--
If your PR closes some issues, please write `Closes #XXXX`
where `XXXX` is the number of the issue you want to fix.
Each issue goes on it's own line.
-->

Refactor coloring/help/errors facility.

### Improvements

1. _Creation_ of of `clap::Error` is now completely separate from actual _emitting_. It cannot fail anymore.
2. Significant code bloat decrease:
   
   Before
   ``` 
    File  .text     Size Crate
    9.4%  56.8% 289.5KiB clap
    4.8%  28.7% 146.2KiB std
    1.9%  11.6%  59.1KiB [Unknown]
    0.2%   0.9%   4.8KiB indexmap
    0.1%   0.3%   1.6KiB textwrap
    0.0%   0.1%     307B probe
    0.0%   0.0%      41B os_str_bytes
   16.6% 100.0% 509.8KiB .text section size, the file size is 3.0MiB
   ```

   After
   ```
    File  .text     Size Crate
    8.6%  53.7% 262.5KiB clap
    5.0%  31.2% 152.5KiB std
    1.9%  12.1%  59.1KiB [Unknown]
    0.2%   1.0%   4.8KiB indexmap
    0.1%   0.3%   1.6KiB textwrap
    0.0%   0.1%     307B probe
    0.0%   0.0%      41B os_str_bytes
   16.1% 100.0% 489.1KiB .text section size, the file size is 3.0MiB
   ```
3. No more our own unicorn reimplementations of `str::split`. I consider this a huge improvement.

### Changes to public API
1. `App::write_[long_]version` is replaced with render version. It was necessary in order to simplify some code. I found that these methods are [very](https://github.com/Rukenshia/saml2aws-auto/blob/1b483059cc8925e98ecde5fc246b10c6088e0882/src/main.rs#L88) [rarely](https://github.com/lise-henry/crowbook/blob/464b323b4c182bcb736ebfc41d197ac0db117bfd/src/bin/helpers.rs#L271) used, and what those people really needed was `String`.

   Actually, in light of `App::get[_long]_version`, I think we could just remove it. @pksunkara Any objections?
2. `Error::cause` and `Error::info` fields are removed. I've never seen them used in practice, and they are responsible for about `10 KiB` of code bloat.
3. `print_help*` and `write_help*` now return `io::Result` instead of `clap::Result`. There's no point to convert them to clap errors.

### Performance regression

I haven't done proper benchmarking yet, but I expect certain regression in error/help rendering. I suppose the regression is negligible because printing is slow enough on it's own, but I'm open to continue working on improving it.

pksunkara commented on src/parse/parser.rs:
```diff
@@ -1653,27 +1644,21 @@ where
-                info: None,
-            },
+        let msg = self.app._render_version(use_long);
+        let mut c = Colorizer::new(false, ColorChoice::Never);
```
```suggestion
        let mut c = Colorizer::new(false, self.app.color());
```

CreepySkeleton replied:
We never color the version, so it doesn't matter.

CreepySkeleton replied:
We never color the version, so it doesn't matter.

pksunkara replied:
Still, if we decide to color in the future. Always passing app's color is better when constructing colorizer.

pksunkara commented on src/parse/errors.rs:
```diff
@@ -375,39 +374,33 @@ pub enum ErrorKind {
 #[derive(Debug)]
 pub struct Error {
-    /// The cause of the error
-    pub cause: String,
```
Why is this removed?

CreepySkeleton replied:
> Error::cause and Error::info fields are removed. I've never seen them used in practice, and they are responsible for about 10 KiB of code bloat.

pksunkara commented:
Something is bugging me about this PR and I can't exactly pinpoint what it is. Which is leading me to procrastinate on reviewing this. 😞

CreepySkeleton commented:
ping @pksunkara

This PR is essentially a blocker for any other change to the error facility because it spans over almost thousand of lines of code over there.

CreepySkeleton commented:
It's not just about bloat, it's about feature creep. This is a golden opportunity to prune unused and redundant stuff, would be a real shame to miss it. 

Actually, this talk was the reason I removed the fields in the first place - I saw some useless code and wanted to trigger the discussion around it in hope somebody would come with a compelling reason not to remove it. Your argument is basically "let's leave it just in case". Did you ever use them? Can you give an example (some code from a real project, not a hello-world)? 

My arguments:

* `cause` is useless because it's `err.to_string().split("\n\n").next().unwrap()`. Literally. I'm not opposed to having a method encapsulating that - which will be removed by LTO if not used by the way, but I don't see any reason in dedicating a separate field for that and even making use of `std::format!` which isn't exactly lightweight.

  I also don't have a tiniest idea of what it can be used for.
* `info` is useless because I fail to see how it can possibly be used. I have talked to a few people and none of them used these fields and none of them could come up with an idea how-to. While there's obviously sampling bias at play, I think "nobody needs it" is good null hypothesis here.

  I'm not so sure on this one in light of #2016 , but that person didn't supply any particular examples either. I believe this is just a miscommunication of sorts and what they want is actually `.to_string()`. Let's continue the discussion there.

pksunkara commented:
`cause` is not useless. I want to use my CLI in 2 different places (one terminal with colors and one where I don't want the colors (ex: chatroom like described in #2016)), I don't want to run the CLI twice. I would prefer both the error messages to exist in just one run.

CreepySkeleton commented:
>  If I want to use my CLI in 2 different places (one terminal with colors and one where I don't want the colors

That piece gives me strong impression that you're mistaking `cause` for `to_string` :) How it actually works:

The `Display` impl doesn't do coloring. Never. This is exactly what you're supposed to use when you want no colors. `Colorizer::print` is what you use if you want colors (`Error::exit` uses just that). We'll need to introduce `pub fn Error::print(stderr: bool)`, but I'd like to postpone it for another PR, this one is already hell of a change.

As I noticed earlier, `cause` is just the first sentence from the rendered error (i.e the message without "did you mean" and "hint" stuff). For what do you think you can use this? And as noted, it can be easily emulated via `err.to_string().split("\n\n").next().unwrap()`. 

I accept your reasoning for `info`, I'll put it back.

pickfire commented on src/build/app/mod.rs:
```diff
@@ -1952,8 +1952,7 @@ impl<'help> App<'help> {
 
-    // used in clap_generate (https://github.com/clap-rs/clap_generate)
-    #[doc(hidden)]
+    /// Some properties of an `App`
```
Should this not be hidden?

pickfire replied:
@CreepySkeleton I think you missed this.

CreepySkeleton replied:
Um, yes, I was going to address the thing in this PR but then figured it would be too off-topic. This chunk somehow slipped through, oops.

pickfire commented:
@CreepySkeleton Since we have the luxury to change stuff. Can we enable color by default, it stays hidden for quite a few crates that depends on clap, I have sent pull requests to some of them and they were surprised by this feature.

CreepySkeleton commented:
Colored help `feature = "color"` will  be on by default in 3.0.

pickfire commented:
> Colored help feature = "color" will be on by default in 3.0.

Yes, the feature is on by default. IIRC even in 2.0 but I mean `AppSettings::ColoredHelp`, not the cargo feature but the clap setting to show colored help by default. Looking at `clap_derive` `basic` example, `cargo run --example basic -- -h` does not show any color.

By the way, I have a comment above that you missed.

CreepySkeleton commented:
@pickfire Could you please create a separate issue about coloring? For the record, I agree. The `Auto, Never, Always` settings set is enough.

pksunkara commented on src/build/app/mod.rs:
```diff
@@ -1572,15 +1573,14 @@ impl<'help> App<'help> {
     /// [`--version` (long)]: ./struct.App.html#method.long_version
-    pub fn write_long_version<W: Write>(&self, w: &mut W) -> ClapResult<()> {
-        self._write_version(w, true).map_err(From::from)
+    pub fn render_long_version(&self) -> String {
```
Hmm.. I understand what we are trying to do here. But can you please comment with a sentence or two on why we don't want `write_` fns anymore?

CreepySkeleton replied:
Damn, I wish I could remember... 

I vaguely recall that was something about weird borrowcheck errors involving the `Write` here. I don't really think there's much point to keep this function around since we now have `get_version`. It doesn't do much anyway.

pksunkara commented on src/parse/errors.rs:
```diff
@@ -440,604 +436,415 @@ impl Error {
     pub(crate) fn value_validation(
-        arg: Option<&Arg>,
-        err: &str,
+        arg: String,
```
Why is this not &Arg? Why is `val` and `err` not `&str`?

CreepySkeleton replied:
Because we don't always have the `Arg`, like in `value_of_t`.

pksunkara commented on src/parse/errors.rs:
```diff
@@ -440,604 +436,415 @@ impl Error {
-        other: Option<O>,
-        usage: U,
+        other: Option<String>,
+        usage: String,
```
Why is this not `&str`?

CreepySkeleton replied:
Because what we need here is `String`, not `&str`. This way the caller can decide how to form the string and reuse the already formed `String` without reallocation on our part. In fact, the caller has `String` most of the time.

pksunkara replied:
I was asking because the `usage` parameter in all the other errors below has `&str` but this one was different. I just wanted to confirm that it was intentional since it's not consistent.

CreepySkeleton replied:
Yes, I was going to make the other methods to use `String` too (I have some ongoing work done locally), but decided to postpone it as quite irrelevant (PR is already larger than necessary). Do you want it in this PR or as a followup?

pksunkara replied:
If it's only `&str` to String and such minor stuff, you can add a new commit to this PR. We will squash after I review and approve the PR.

CreepySkeleton commented on src/build/app/mod.rs:
```diff
@@ -1663,18 +1663,15 @@ impl<'help> App<'help> {
-    /// [`--help` (long)]: ./struct.Arg.html#method.long_about
-    pub fn print_help(&mut self) -> ClapResult<()> {
+    /// [`-h` (short)]: ./struct.Arg.html#method.help
+    /// [`--help` (long)]: ./struct.Arg.html#method.long_help
```
And this is exactly why I said rebasing PRs of this size is _painful_. Some artifacts slip through almost inevitably, and I've done it more than once.

CreepySkeleton commented on src/output/fmt.rs:
```diff
@@ -20,103 +21,115 @@ fn is_a_tty(stderr: bool) -> bool {
-            when
+    /// Color-aware printing. Attempts to draw colors whenever possible.
+    #[cfg(feature = "color")]
+    fn print_colored(&self) -> io::Result<()> {
```
I don't understand what you're talking about. I don't understand your motivation. And there's already the `print` function dozen lines below.

pksunkara replied:
So, you have the following currently:

```rust
#[cfg(feature = "color")]
fn print_colored(&self) -> io::Result<()> {
    // color print logic
}

#[inline]
pub(crate) fn print(&self) -> io::Result<()> {
    #[cfg(feature = "color")]
    {
        self.print_colored()
    }

    #[cfg(not(feature = "color"))]
    {
        // non-color print logic
    }
}
```

Now, the whole thing can be simplified much more easily and made much more readable:

```rust
#[cfg(feature = "color")]
pub(crate) fn print(&self) -> io::Result<()> {
    // color print logic
}

#[cfg(not(feature = "color"))]
pub(crate) fn print(&self) -> io::Result<()> {
    // non-color print logic
}
```

CreepySkeleton replied:
:+1

CreepySkeleton commented on src/parse/errors.rs:
```diff
@@ -375,39 +374,36 @@ pub enum ErrorKind {
-    pub info: Option<Vec<String>>,
+    /// Additional information depending on the error kind, like values and argument names.
+    /// Useful when you want to render an error of your own.
+    pub info: Vec<String>,
```
Which info, exactly? The meaning is the same as empty vec.

CreepySkeleton commented on src/output/help.rs:
```diff
@@ -860,242 +842,161 @@ impl<'help, 'app, 'parser, 'writer> Help<'help, 'app, 'parser, 'writer> {
+                            self.none("\n")?;
+                        }
+                    }
+                    "long-about" => {
```
Another artifact.

CreepySkeleton commented on src/output/fmt.rs:
```diff
@@ -20,103 +21,115 @@ fn is_a_tty(stderr: bool) -> bool {
-        Ok(())
+    /// Chooses printing method depending on features.
+    /// (Display impl vs Colorizer::print_colored)
+    #[inline]
```
Maybe not, but I'd very much like to address inline strategy in a separate PR. One thing at a time. (Inlineness is not an API property and adding/removing it isn't breaking).

CreepySkeleton commented on src/build/app/mod.rs:
````diff
@@ -1786,15 +1787,14 @@ impl<'help> App<'help> {
     /// ```rust
````
> Also, the header for Examples is not ###

Yes, and it's been like that all the time. I'll fix this instance, but I'm leaving a through proofread for another PR.

pksunkara reviewed (changes requested):
Also, can you please squash the commits once the PR is ready. I don't mind if it's just one large commit.

pksunkara reviewed (changes requested):
You misrebased the tests/help.rs file.

pksunkara commented:
Your tests are failing because new stable is released. I fixed the UI tests in #2114

pksunkara reviewed (approved):
Can you please squash now?

pksunkara commented:
Once squashed, you can do `bors r=pksunkara` to merge. 👍

CreepySkeleton commented:
bors r=pksunkara
